text/0074-stateful-multisig-pallet.md

```diff
@@ -0,0 +1,293 @@
+
+All functions have detailed rustdoc in [PR#3300](https://github.com/paritytech/polkadot-sdk/pull/3300). Here is a brief overview of the functions:
+
+* `create_multisig` - Create a multisig account with a given threshold and initial owners. (Needs Deposit)
+* `start_proposal` - Start a multisig proposal. (Needs Deposit)
+* `approve` - Approve a multisig proposal.
+* `revoke` - Revoke a multisig approval from an existing proposal.
+* `execute_proposal` - Execute a multisig proposal. (Releases Deposit)
+* `cancel_own_proposal` - Cancel a multisig proposal started by the caller in case no other owners approved it yet. (Releases Deposit)
+
+Note: Next functions need to be called from the multisig account itself. Deposits are reserved from the multisig account as well.
+
+* `add_owner` - Add a new owner to a multisig account. (Needs Deposit)
+* `remove_owner` - Remove an owner from a multisig account. (Releases Deposit)
+* `set_threshold` - Change the threshold of a multisig account.
+* `cancel_proposal` - Cancel a multisig proposal. (Releases Deposit)
+* `delete_multisig` - Delete a multisig account. (Releases Deposit)
+
+### Storage/State
+
+* Use 2 main storage maps to store mutlisig accounts and proposals.
+
+```rust
+#[pallet::storage]
+  pub type MultisigAccount<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, MultisigAccountDetails<T>>;
+
+/// The set of open multisig proposals. A proposal is uniquely identified by the multisig account and the call hash.
+/// (maybe a nonce as well in the future)
+#[pallet::storage]
+pub type PendingProposals<T: Config> = StorageDoubleMap<
+    _,
+    Twox64Concat,
+    T::AccountId, // Multisig Account
+    Blake2_128Concat,
+    T::Hash, // Call Hash
+    MultisigProposal<T>,
+>;
+```
+
+As for the values:
+
+```rust
+pub struct MultisigAccountDetails<T: Config> {
+	/// The owners of the multisig account. This is a BoundedBTreeSet to ensure faster operations (add, remove).
+	/// As well as lookups and faster set operations to ensure approvers is always a subset from owners. (e.g. in case of removal of an owner during an active proposal)
+	pub owners: BoundedBTreeSet<T::AccountId, T::MaxSignatories>,
+	/// The threshold of approvers required for the multisig account to be able to execute a call.
+	pub threshold: u32,
+	pub creator: T::AccountId,
+	pub deposit: BalanceOf<T>,
+}
+```
+
+```rust
+pub struct MultisigProposal<T: Config> {
+    /// Proposal creator.
+    pub creator: T::AccountId,
+    pub creation_deposit: BalanceOf<T>,
+    /// The extrinsic when the multisig operation was opened.
+    pub when: Timepoint<BlockNumberFor<T>>,
+    /// The approvers achieved so far, including the depositor.
+    /// The approvers are stored in a BoundedBTreeSet to ensure faster lookup and operations (approve, revoke).
+    /// It's also bounded to ensure that the size don't go over the required limit by the Runtime.
+    pub approvers: BoundedBTreeSet<T::AccountId, T::MaxSignatories>,
```

Comment:
do we not need to store rejectors? e.g. for 2/4 multisig, the proposal can be removed if 3 signers rejects it

Reply:
The initial design had `cancel_proposal` as a separate multisig operation. Meaning that to cancel the proposal they'd need another proposal to do so. This is not optimal and I agree with you suggestion. After reaching rejections threshold, any signer can call `cancel_proposal` like `execute_proposal`. Working on updating the RFC. Thanks!